System.Net.CookieParser to Support Cookie String with Trailing Semicolon

shmao opened the pull request:
The issue was that if a cookie string has trailing semicolon, `System.Net.CookieParser.GetString()` returns empty string. The issue was found on UAP/UAPAOT runs as `System.Net.CookieParser.GetString()` was used on those platforms.

We would also need to port the fix to UWP6.0 branch.

Fix #22877

davidsh commented on src/System.Net.Primitives/tests/UnitTests/CookieInternalTest.cs:
```diff
@@ -41,5 +41,67 @@ public void DomainImplicit_AddCookieWithExplicitDomain_DomainImplicitFalse()
             Assert.False(cookies[CookieName].DomainImplicit);
         }
+
+        [Fact]
```
Considering that all these tests are testing the same operations but just with different inputs, you should collapse all of these into a  singleTheory and use inlinedata attributes to pass in the various different inputs you are testing.

And in terms of the test method naming convention that we use in all our tests, we use a 3-part naming, such as:

><methodOrPropertyUnderTest>_<conditions>_<expectedResult>

So, perhaps the theory can be named "CookieParserGetString_SetCookieHeaderValue_Success" etc.

davidsh commented:
Strictly speaking, these unit tests are only testing the change in the src/Common/src/System/Net/CookieParser.cs file.  But since the change is identical in the uap copy of it, it should be ok.  We still need to add functional tests but we can do that later considering the urgency of this fix for UWP6.0 milestone.

davidsh commented:
Please also run Outerloop tests as well.  These commands will launch those tests.
@dotnet-bot Test Outerloop Linux x64 Release Build
@dotnet-bot Test Outerloop UWP CoreCLR x64 Debug Build
@dotnet-bot Test Windows x64 Debug Build

shmao commented:
@davidsh Thanks for reviewing the PR. I've changed the [Fact]s to [Theory]s per your suggestion.

@dotnet-bot Test Outerloop Linux x64 Release Build
@dotnet-bot Test Outerloop UWP CoreCLR x64 Debug Build
@dotnet-bot Test Windows x64 Debug Build

davidsh commented:
@dotnet-bot Test Outerloop Linux x64 Release Build
@dotnet-bot Test Outerloop UWP CoreCLR x64 Debug Build
@dotnet-bot Test Windows x64 Debug Build

stephentoub commented on src/System.Net.Primitives/tests/UnitTests/CookieInternalTest.cs:
```diff
@@ -41,5 +41,29 @@ public void DomainImplicit_AddCookieWithExplicitDomain_DomainImplicitFalse()
+        public void CookieParserGet_SetCookieHeaderValue_Success(string name, string value, string cookieString)
+        {
+            var parser = new CookieParser(cookieString);
+            Cookie cookie= parser.Get();
```
Nit: space before =

stephentoub commented on src/System.Net.Primitives/tests/UnitTests/CookieInternalTest.cs:
```diff
@@ -41,5 +41,29 @@ public void DomainImplicit_AddCookieWithExplicitDomain_DomainImplicitFalse()
+        [Theory]
+        [InlineData("cookie_name", "cookie_value", "cookie_name=cookie_value")]
+        [InlineData("cookie_name", "cookie_value", "cookie_name=cookie_value;")]
+        public void CookieParserGet_SetCookieHeaderValue_Success(string name, string value, string cookieString)
```
Nit: consider reordering and renaming the args as:
```
(string cookieString, string expectedName, string expectedValue)
```

stephentoub commented on src/System.Net.Primitives/tests/UnitTests/CookieInternalTest.cs:
```diff
@@ -41,5 +41,29 @@ public void DomainImplicit_AddCookieWithExplicitDomain_DomainImplicitFalse()
+
+        [Theory]
+        [InlineData("cookie_name", "cookie_value", "cookie_name=cookie_value")]
+        [InlineData("cookie_name", "cookie_value", "cookie_name=cookie_value;")]
```
Do we have / need a test where there are multiple cookies and it ends in a semicolon?

shmao commented:
@dotnet-bot Test Outerloop Linux x64 Release Build
@dotnet-bot Test Outerloop UWP CoreCLR x64 Debug Build
@dotnet-bot Test Windows x64 Debug Build

shmao commented:
I added more tests per @stephentoub 's suggestion. One of the tests, `CookieParserGet_SetCookieHeaderValue_Success`, failed. I talked to @davidsh offline. We thought it's a different issue. I opened #22925 for tracking it.

For the record, the test mentioned above failed before/after this fix. The test is not directly related to the fix. I added it for extra coverage for `CookieParser`.

shmao commented:
Test `System.Net.Http.Functional.Tests.HttpClientHandler_SslProtocols_Test` failed with the error below,

```
GetAsync_UnsupportedSSLVersion_Throws(name: \"SSLv2\", url: \"https://www.ssllabs.com:10200/\")

Message :
Assert.Throws() Failure\nExpected: typeof(System.Net.Http.HttpRequestException)\nActual:   typeof(System.Threading.Tasks.TaskCanceledException): A task was canceled.
Stack Trace :
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at System.Net.Http.HttpClient.<FinishSendAsyncBuffered>d__58.MoveNext() in /mnt/j/workspace/dotnet_corefx/master/linux-TGroup_netcoreapp+CGroup_Release+AGroup_x64+TestOuter_true_prtest/src/System.Net.Http/src/System/Net/Http/HttpClient.cs:line 464
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
```

The test passed on my machine. And it also passed in the previous iterations. The latest iteration just changed test code.

Is this a known issue? Can I ignore this failure?

@davidsh @geoffkizer 
I

davidsh commented:
>Is this a known issue? Can I ignore this failure?

Yes, ignore this for your PR.

shmao commented:
@davidsh sounds good, Thanks. Merging the PR.
